### DISCUSSION TITLE
YAML Editor Question

### DISCUSSION THREAD

**Discussion by jl-beast** [kubeapps/kubeapps#6727] [Q&A]
I have the following setup. I have a Chart with a Chart.yaml and in the folder with the Chart there exists 4 other yaml files:

- deploy-cloud-values.yaml
- deploy-values.yaml
- local-values.yaml
- values.yaml

The Deploy ones are used for doing test runs with integration tests and other such. Local-Values is used for configuring it to run given that one does not have specific container infrastructure locally. And the regular values.yaml is the base for them all.

When I boot up, the default values in the YAML editor is deploy-cloud-values.yaml, whereas I was expecting the values.yaml which has lovely comments to help with editing. 
- Is this intended behavior? 
- And if it is, is there a way such that I can switch between these values.yaml files in the yaml editor? 

I am using 13.0.1 with App version of 2.8.0

I also tried with a 2.6.0 version, which shows values.yaml.

**Comment by absoludity** [comment]
Hi @jl-beast . If Kubeapps detects more than one extra values, it should present you with a drop-down to select the additional file that you want to use. If Kubeapps detects only one extra values file, it will display that without any option currently (though see #6564).

Regarding you lovely comments to help with editing: the intention here is that if you as the chart author provide an extra values file (or multiple), that they will display exactly what you want users to see (ie. it may duplicate important options from the values.yaml, but would normally be a small subset that you think are most relevant, or may have overrides that you want), and should be commented for your users in the same way. It is only an overlay of the original values.yaml (as in, it'll end up running with `helm install -v values.yaml -v values-custom-file.yaml`).

Note that in our code, we're checking for [additional files that have the prefiix "values-"](https://github.com/vmware-tanzu/kubeapps/blob/39a84dd3b661b83846a869143e825aa3aeb22c2d/cmd/asset-syncer/server/utils.go#L1055-L1068), which wouldn't match any of your extra files. That is, Kubeapps *should* be picking up and matching files such as:

-  values-deploy-cloud.yaml
- values-deploy.yaml
- values-local.yaml

in addition to the normal values.yaml, but not the ones you listed. So I'm surprised and confused that you're seeing your `deploy-could-values.yaml`, which [definitely doesn't match](https://play.golang.com/p/RCOEuP9mdbJ). Let me know if you accidentally used different file names here, or otherwise please turn this into an issue and we'll look into why it's picking up those when it shouldn't.

**Reply by jl-beast** [reply]
@absoludity 

I renamed my files and that worked, but I noticed a couple of things. 

1. In the visual editor, I have a variable called image.tag, which is used to specify the image.tag and we have it set to a release on the regular values.yaml. When I switch to values-deploy.yaml (new name), It does not populate with the actual value that it will deploy with.
2. The Yaml Editor does not replace values as a diff showing the exact values.yaml that is being deployed. Instead it shows the actual values.yaml, which is fine, but it would be much more preferable to see exactly what you are deploying with values-deploy.yaml being put over as a diff similar to entering values manually.
3. I was unable to find exact documentation that this feature exists and I think it is rather important as one of the niceties of kubeapps is that you can generally get some non-engineers to do deployments. 

Really appreciative of the fast turn-around, enjoying the product :)

**Reply by absoludity** [reply]
Great - glad that was the (main) issue. Let me address your points:

1. Correct - it will show the contents of the file you've selected (`values-deploy.yaml`), as whatever value you specify there is going to overwrite what's in the values.yaml (though I'm not sure of Helm's behaviour if it's blank in `values-deploy.yaml` - would need to check). If your intention is to have a simplified values for users to interact with (in `values-deploy.yaml`) **and** you want to present that value with a default when users use that values file, then you'll need to populate the value with that default.
2. If I'm understanding correctly, your point is that the diff shows the diff of changes that the user has made **with the selected values file** rather than the diff of what they've changed to what's being deployed? Though I don't think that's the case (in that any value specified in the selected `values-deploy.yaml` will overwrite what's in the standard `values.yaml`, so the diff should be correct even for the user changes against the standard `values.yaml`? (though maybe not with blank values, depending what Helm does with those). Sorry, I'm not sure I'm following this one. Maybe a screenshot with annotations would help?
3. Yes, ah-hem, really sorry about that. It was a new feature that we added write when were then low on resources and it's been hard to prioritise. But please do leave a comment on #5987 saying it's important to you and we might get to put some time aside :)

Glad you're enjoying Kubeapps!
